src/decoder/lattice-incremental-decoder.h

```diff
@@ -0,0 +1,655 @@
+   FST == StdFst and it notices that the actual FST type is
+   fst::VectorFst<fst::StdArc> or fst::ConstFst<fst::StdArc>, the decoder object
+   will internally cast itself to one that is templated on those more specific
+   types; this is an optimization for speed.
+ */
+template <typename FST, typename Token = decoder::StdToken>
+class LatticeIncrementalDecoderTpl {
+ public:
+  using Arc = typename FST::Arc;
+  using Label = typename Arc::Label;
+  using StateId = typename Arc::StateId;
+  using Weight = typename Arc::Weight;
+  using ForwardLinkT = decoder::ForwardLink<Token>;
+
+  // Instantiate this class once for each thing you have to decode.
+  // This version of the constructor does not take ownership of
+  // 'fst'.
+  LatticeIncrementalDecoderTpl(const FST &fst, const TransitionModel &trans_model,
+                               const LatticeIncrementalDecoderConfig &config);
+
+  // This version of the constructor takes ownership of the fst, and will delete
+  // it when this object is destroyed.
+  LatticeIncrementalDecoderTpl(const LatticeIncrementalDecoderConfig &config,
+                               FST *fst, const TransitionModel &trans_model);
+
+  void SetOptions(const LatticeIncrementalDecoderConfig &config) { config_ = config; }
+
+  const LatticeIncrementalDecoderConfig &GetOptions() const { return config_; }
+
+  ~LatticeIncrementalDecoderTpl();
+
+  /// An example of how to do decoding together with incremental
+  /// determinization. It decodes until there are no more frames left in the
+  /// "decodable" object. Note, this may block waiting for input
+  /// if the "decodable" object blocks.
+  /// In this example, config_.determinize_delay, config_.determinize_period
+  /// and config_.determinize_max_active are used to determine the time to
+  /// call GetLattice().
+  /// Users may do it in their own ways by calling
+  /// AdvanceDecoding() and GetLattice(). So the logic for deciding
+  /// when we get the lattice would be driven by the user.
+  /// The function returns true if any kind
+  /// of traceback is available (not necessarily from a final state).
+  bool Decode(DecodableInterface *decodable);
+
+  /// says whether a final-state was active on the last frame.  If it was not, the
+  /// lattice (or traceback) will end with states that are not final-states.
+  bool ReachedFinal() const {
+    return FinalRelativeCost() != std::numeric_limits<BaseFloat>::infinity();
+  }
+
+  /// Outputs an FST corresponding to the single best path through the lattice.
+  /// Returns true if result is nonempty (using the return status is deprecated,
+  /// it will become void).  If "use_final_probs" is true AND we reached the
+  /// final-state of the graph then it will include those as final-probs, else
+  /// it will treat all final-probs as one.
+  bool GetBestPath(Lattice *ofst, bool use_final_probs = true);
+
+  /**
+    The following function is specifically designed for incremental
```

Comment:
It's good that you are making an effort to explain things, but we should separate explanatin of the interace and external behavior, from explanation of the algorithm.  It may be better to just refer to the paper for explanation of the algorithm.